### ISSUE TITLE
KhalimskyCell and KhalimskyPreCell: are both types needed?

### ISSUE THREAD

**Issue by phcerdan** [DGtal-team/DGtal#1533]
I get the difference between KSpace and KPreSpace, but I don't see the difference between `KCell` and `KPreCell`.
Basically, a `KCell` `has-a` `KPreCell`, but they do share the same interface.

Do we need both types? What am I missing? @rolanddenis @JacquesOlivierLachaud

**Comment by rolanddenis** [comment]
Yes but automatic conversion is disabled from `KPreCell` to `KCell` and `KhalimskySpace` only accepts `KCell` for "critical" functionalities. A `KPreCell` can only be converted to `KCell` through specific methods of `KhalimskySpace`. The idea was to ensure (at least in Debug mode) that `KhalimskyCell` are valid (ie inside the bounds) while being able to do some nasty manipulations on coordinates of `KhalimskyPreCell` (direct modification of `KhalimskyCell`'s coordinates isn't possible). It has been useful to fix many bugs but there was another reason for another Khalimsky space implementation that @JacquesOlivierLachaud want to add (don't remember exactly).

**Comment by rolanddenis** [comment]
I think there was also another reason : for periodic boundaries, coordinates must be kept inside the domain where moving through the boundaries. There is two way of doing this :
1. applying the modulo each time the coordinates are read,
2. applying the modulo only when the coordinates are modified.

For performance reasons, the second option is better but it implies to control coordinate modifications and thus forcing user to use dedicated methods of `KhalimskySpaceND`.

**Comment by JacquesOlivierLachaud** [comment]
Roland is right. At the origin, I wished to encapsulate "movements" of cells through methods in KhalimskySpaceND, so that the user is not dependent on the chosen representation of a cell.
What I had in mind was my "optimized" model of Khalimsky space, where every cell is coded entirely in one int (or int64).
The idea was that, in a bounded space, we can dedicate a few bits for each coordinate, a few bits for the topology and a bit for the orientation. Once this is done, we can code all basic operations with integer bitwise operations and LUT.
(see #94 ).
Roland did a very nice job when implementing toric spaces to keep this requirement.
Unfortunately I am not sure that this original nice idea was sufficiently respected in all the produced code.
I remember some place where cell coordinates are accessed and modified directly (for instance in the viewers at a time).
And finally, for this nice idea to be really a speed up, you have to provide the user with a DigitalSet model that  encodes a characteristic function, with one bit per possible cell. This was done in my former library ImaGene with nice speed-up (from 3 to 10 times faster). If you only encode the cells (and keep std::set or std::unordered_set), the speed-up was more 20-30% (with a few experiments I made at the time).

**Comment by phcerdan** [comment]
Thanks @rolanddenis @JacquesOlivierLachaud for the clarifications, they make sense.

@JacquesOlivierLachaud I really liked the idea of FastKSpace after reading your article and some parts of your branch (I read the `Code` part, generated by joining `BitFields`, where a field can encode: coordinate, topology, orientation. It might be worth the effort to re-implement the idea on a `DigitalSetEncoded.h` or similar. 3x-10x speed up is quite significant!
I am guessing for that to work we will need another `PreCell` type, but instead of the `coordinates` member being a `Point`, being a `Code` (int).
I would love to help with this, maybe next year I find some time.

**Comment by JacquesOlivierLachaud** [comment]
This is exactly the idea ! I would be glad to help you/discuss with you on this topic when you have the time.
It's unlikely I will do it in the next months (since I already postponed it for quite a number of years :-D )

**Comment by rolanddenis** [comment]
@phcerdan A new `PreCell` should not be needed since it is dedicated to unbounded spaces (that is incompatible with encoding) but instead just a `FastCell` that owns a `Code` and can be converted to a `PreCell` (but not owning it).

**Comment by phcerdan** [comment]
I see, thanks @rolanddenis.
We would need:
- KhalimskySpaceEncoded.h
- DigitalSetEncoded.h
- CellEncoded.h or `FastCell`
